src/coreclr/tools/aot/ILCompiler.ReadyToRun/JitInterface/CorInfoImpl.ReadyToRun.cs

```diff
@@ -2327,17 +2308,6 @@ private void EncodeFieldBaseOffset(FieldDesc field, CORINFO_FIELD_INFO* pResult,
                 }
                 // ENCODE_NONE
             }
-            else if (HasLayoutMetadata(pMT))
-            {
-                PreventRecursiveFieldInlinesOutsideVersionBubble(field, callerMethod);
-
-                // We won't try to be smart for classes with layout.
```

Comment:
Actually, the old crossgen has a special case for this with this comment that kicks in for non-value types (different from what was here before that kicked in for value types).

Do we want to match it? Or are we confident that we have extensive test coverage for all non-value types with layout corner cases?

Also, once we add this back, can at least some of the remaining changes be simplified?

Reply:
According to my understanding of the CoreCLR code, the layout check is based on the following condition:

https://github.com/dotnet/runtime/blob/207b03a073107e63df0b67056dd596226065651e/src/coreclr/vm/methodtablebuilder.cpp#L11924

It seems to me that this means that the conditional statement you originally pointed out should basically stay, the problematic bit is the method HasLayoutMetadata itself - we're certainly trying to match the CoreCLR runtime behavior, in this particular case we probably just missed the fact that the CoreRT version of HasLayoutMetadata is quite different than its CoreCLR runtime counterpart method

https://github.com/dotnet/runtime/blob/207b03a073107e63df0b67056dd596226065651e/src/coreclr/vm/methodtablebuilder.cpp#L11746

I'll look into putting HasLayoutMetadata back and modifying it based on the CoreCLR runtime version. Other than that, can you please be more specific regarding the simplifications you're suggesting? While I do agree that some of the runtime logic is complicated and somewhat arbitrary, I'm worried that unsettling it at this point might cause more harm than good.

Reply:
I think the equivalent of the HasLayoutMetadata should be just `type.IsSequentialLayout || type.IsExplicitLayout`.

Yes, the runtime logic is complicated. We have done multiple rounds of trying to replicate it, and we are still finding mismatches. It is safe bet that this is not the last round of fixes. I think we need to make it simpler to get confident that it matches. I will do some ad-hoc testing to try to find more situations where it does not match.

Reply:
Thanks Jan for your additional feedback. I have updated the runtime check based on your suggestion. Once the change passes basic testing, I'll also trigger another round of Pri1 testing that previously uncovered most of these inconsistencies. It would be certainly awesome if you were willing to use your vast expertise in this area to devise additional tests exercising various obscure corner cases we may have previously missed.